File: library/src/main/java/name/zeno/android/widget/indicator/TextPageIndicator.java
package name.zeno.android.widget.indicator;

import android.content.Context;
import android.support.v4.view.ViewPager;
import android.support.v7.widget.AppCompatTextView;
import android.util.AttributeSet;

import name.zeno.android.widget.PageIndicator;

/**
 * Create Date: 16/7/4
 *
 * @author 陈治谋 (513500085@qq.com)
 */
public class TextPageIndicator extends AppCompatTextView implements PageIndicator
{
  private ViewPager pager;

  public TextPageIndicator(Context context)
  {
    this(context, null);
  }

  public TextPageIndicator(Context context, AttributeSet attrs)
  {
    this(context, attrs, 0);
  }

  public TextPageIndicator(Context context, AttributeSet attrs, int defStyleAttr)
  {
    super(context, attrs, defStyleAttr);
  }

  @Override public void notifyDataSetChanged()
  {
    invalidate();
  }

  @Override public void setViewPager(ViewPager view)
  {
    setViewPager(view, 0);
  }

  @Override public void setViewPager(ViewPager view, int initialPosition)
  {
    pager = view;
    view.addOnPageChangeListener(this);
  }

  @Deprecated
  @Override public void setCurrentItem(int item)
  {
    pager.setCurrentItem(item);
  }

  @Override public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels)
  {
    float scale = 0.5F + (float) Math.abs(positionOffset - 0.5);
    setAlpha(scale);

    int currP = pager.getCurrentItem();
    int count = pager.getAdapter().getCount();
    if (currP == position) {
      currP = currP + count - 1 % count;
    }
    onPageSelected(positionOffset <= 0.5 ? currP : position);
  }

  @Override public void onPageSelected(int position)
  {
    int count = pager.getAdapter().getCount();
    position = position % count;
    setText(pager.getAdapter().getPageTitle(position));
  }

  @Override public void onPageScrollStateChanged(int state)
  {

  }
}

File: library/src/main/java/name/zeno/android/widget/indicator/TextPageIndicator.kt
package name.zeno.android.widget.indicator

import android.content.Context
import android.support.v4.view.ViewPager
import android.support.v7.widget.AppCompatTextView
import android.util.AttributeSet

import name.zeno.android.widget.PageIndicator

/**
 * @author 陈治谋 (513500085@qq.com)
 * @since  16/7/4
 */
class TextPageIndicator @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : AppCompatTextView(context, attrs, defStyleAttr), PageIndicator<ViewPager> {
  private var pager: ViewPager? = null

  override fun notifyDataSetChanged() {
    invalidate()
  }

  override fun setViewPager(view: ViewPager) {
    setViewPager(view, 0)
  }

  override fun setViewPager(view: ViewPager, initialPosition: Int) {
    pager = view
    view.addOnPageChangeListener(this)
  }

  @Deprecated("")
  override fun setCurrentItem(item: Int) {
    pager!!.currentItem = item
  }

  override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
    val scale = 0.5f + Math.abs(positionOffset - 0.5).toFloat()
    alpha = scale

    var currP = pager!!.currentItem
    val count = pager!!.adapter!!.count
    if (currP == position) {
      currP = currP + count - 1 % count
    }
    onPageSelected(if (positionOffset <= 0.5) currP else position)
  }

  override fun onPageSelected(position: Int) {
    val count = pager?.adapter?.count ?: return
    val p = position % count
    text = pager?.adapter?.getPageTitle(p)
  }

  override fun onPageScrollStateChanged(state: Int) {

  }
}
